x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -415,13 +414,13 @@ GetUserPrivilegesResponse buildUserPrivilegesResponseObject(Role userRole) {
         final Set<String> cluster = new TreeSet<>();
         // But we don't have a meaningful ordering for objects like ConfigurableClusterPrivilege, so the tests work with "random" ordering
         final Set<ConfigurableClusterPrivilege> conditionalCluster = new HashSet<>();
-        for (Tuple<ClusterPrivilege, ConfigurableClusterPrivilege> tup : userRole.cluster().privileges()) {
-            if (tup.v2() == null) {
-                if (ClusterPrivilege.NONE.equals(tup.v1()) == false) {
-                    cluster.addAll(tup.v1().name());
-                }
+        for (ClusterPrivilege privilege : userRole.cluster().privileges()) {
+            if (privilege instanceof ActionClusterPrivilege) {
+                cluster.add(((ActionClusterPrivilege) privilege).name());
```

Comment:
Probably better to code to thee interface `Nameable`. Shouldn't we outsource building the list of `Nameable`s to the `ClusterPermission`?

Reply:
This is the only place in core where we iterate over `cluster().privileges()` and we do it to separate the list. So the complete list is not useful, only as parts. I was suggesting we expose `cluster().nameablePrivileges()` and `cluster().configurablePrivileges()`. The issue is probably that the `ClusterPrivilege` parent interface is too slim to be useful in general; the Polymorphism is not useful, so we might as well not use it. 
I don't feel strongly about it, since this method is effectively a builder for `GetUserPrivilegesResponse` (though it does not look like a proper Builder, which is another thing I might change).
Just a suggestion.

Reply:
I don't think we should add these methods. I don't think `Permission` cares (or should care) about the privilege types, and there is only place where we call this method and split by type, so I think it is fine.

In practice, what we have is an _Algebraic Data Type_ of `Named | Configurable`, but since Java has no real support to model ADTs, I think the current code is fine.